secrets/provider/kubernetes/kubernetes_test.go

```diff
@@ -0,0 +1,315 @@
+			Annotations: map[string]string{
+				"hello": "world",
+				"fred":  "mary",
+			},
+		},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(cleanupsForCreate, gc.HasLen, 1)
+	c.Assert(rb.Name, gc.Equals, rbName)
+
+	// get the role binding to check it was created correctly
+	res, err = s.k8sclient.client.RbacV1().RoleBindings(s.k8sclient.namespace).Get(ctx, rbName, v1.GetOptions{})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(res.Name, gc.Equals, rbName)
+	c.Assert(res.Labels, gc.DeepEquals, map[string]string{"app.kubernetes.io/managed-by": "juju", "app.kubernetes.io/name": "app-name"})
+	c.Assert(res.Annotations, gc.DeepEquals, map[string]string{"hello": "world", "fred": "mary"})
+
+	// ensure role binding should get the current role binding with no cleanups if it already exists
+	rb, cleanupsForUpdate, err := s.k8sclient.ensureRoleBinding(ctx, &rbacv1.RoleBinding{
+		ObjectMeta: v1.ObjectMeta{
+			Name: rbName,
+		},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(rb.Name, gc.Equals, rbName)
+	c.Assert(cleanupsForUpdate, gc.HasLen, 0)
+	c.Assert(rb.Labels, gc.DeepEquals, map[string]string{"app.kubernetes.io/managed-by": "juju", "app.kubernetes.io/name": "app-name"})
+	c.Assert(rb.Annotations, gc.DeepEquals, map[string]string{"hello": "world", "fred": "mary"})
+
+	// Run cleanups and verify resources are removed
+	for _, fn := range cleanupsForCreate {
+		fn()
+	}
+	_, err = s.k8sclient.client.RbacV1().RoleBindings(s.k8sclient.namespace).Get(ctx, rbName, v1.GetOptions{})
+	c.Assert(k8serrors.IsNotFound(err), gc.Equals, true)
+}
+
+func (s *kubernetesSuite) TestUpdateClusterRole(c *gc.C) {
+	ctx := context.Background()
+
+	// assert that the cluster role update fails if the cluster role does not exist
+	clusterRoleUpdate := &rbacv1.ClusterRole{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "cluster-role-name",
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{"group1"},
+				Resources: []string{"secrets"},
+				Verbs:     []string{"get", "list"},
+			},
+		},
+	}
+	cr, err := s.k8sclient.updateClusterRole(ctx, clusterRoleUpdate)
+	c.Assert(errors.Is(err, errors.NotFound), gc.Equals, true)
+	c.Assert(cr, gc.IsNil)
+
+	// create the cluster role
+	clusterRoleCreated := &rbacv1.ClusterRole{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "cluster-role-name",
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups:     []string{"group3"},
+				Resources:     []string{"pods"},
+				Verbs:         []string{"get", "list", "watch"},
+				ResourceNames: []string{"pod1", "pod2"},
+			},
+		},
+	}
+	created, err := s.k8sclient.client.RbacV1().ClusterRoles().Create(ctx, clusterRoleCreated, v1.CreateOptions{
+		FieldManager: resources.JujuFieldManager,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(created.Name, gc.Equals, "cluster-role-name")
+	c.Assert(created.Rules, gc.DeepEquals, clusterRoleCreated.Rules)
+
+	// update the cluster role and assert that the rules are updated correctly
```

Comment:
nit: start comments with capital letter and end with "."

Reply:
refactored w thanks!